File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def solution(s):
    answer = 0
    
    queue = deque(s)
    length = len(queue)
    cnt = len(queue)
    q = deque()
    
    
    while cnt > 0:
        stack = []
        
        for item in queue:
            
            if len(stack) == 0:
                stack.append(item)
            else:
                if stack[-1] == "{" and item == "}":
                        stack.pop()
                elif stack[-1] == "(" and item == ")":
                        stack.pop()
                elif stack[-1] == "[" and item == "]":
                        stack.pop()
                else:
                    stack.append(item)
        
        if len(stack) == 0:
            answer += 1
            
        element = queue.popleft()
        queue.append(element)
        cnt -= 1
        
    
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    private static ArrayList<Character> list;
    private static ArrayDeque<Character> stack;
    private static int n;
    
    private static boolean isPossible() {
        stack = new ArrayDeque<>();
        
        for (int i = 0; i < n; i++) {
            char ch = list.get(i);
            
            if (ch == ']' || ch == ')' || ch == '}') {
                if (stack.isEmpty()) return false;
                else {
                    char c = stack.peek();
                    if (ch == ']' && c == '[') stack.pop();
                    else if (ch == ')' && c == '(') stack.pop();
                    else if (ch == '}' && c == '{') stack.pop();
                }
            }
            else stack.push(ch);
        }
        
        return stack.size() == 0 ? true : false;
    }
    public int solution(String s) {
        /*
        s 최대 길이는 1,000이기에 O(N^2) 초과하는 시간복잡도는 불가
        
        [풀이]
        stack 선언해서 s를 전부 push
        pop하면서 올바른 괄호 여부 확인
        
        한 번 확인 끝났으면 앞에 인덱스 요소 제거해서 뒤로 add
        */
        
        int answer = 0;
        n = s.length();
        list = new ArrayList<>();
        
        for (int i = 0; i < n; i++) { // n
            list.add(s.charAt(i));
        }
        
        for (int i = 0 ; i < n; i++) { // n^2
            char ch = list.get(0);
            
            if (isPossible()) answer++;
            list.remove(0);
            list.add(ch);
        }
        
        return answer;
    }
}